Feature: mark price

tailingchen opened the pull request:
**Background**

Currently, in PERP V2, we use the index TWAP as the basis for margin calculation and liquidations. The reason to use index TWAP is it’s hard to make price manipulation on index TWAP. On the other hands, it brings the risk to protocol. At most of time, index TWAP can not response the actual market status. Protocol will get high risk once the spread between index TWAP and market price is getting bigger. The problems are

1. Trader accounts that should be liquidated are not liquidated due to the delayed reaction of index TWAP to current market conditions.
2. When the index TWAP is too different from the market price, the liquidation capacity will decrease. Since PERP v2 uses index TWAP as the close price for liquidation, the liquidators have no incentives to take over the liquidated positions if the discount taking over price is not better than market price.


To reduce potential risks mentioned above, we propose a new mark price calculation which will be used for margin calculation and liquidation.

The `Mark Price` is the median of three prices as below.
1. current market price
2. market twap with 30 mins
3. index price + premium with 15 mins

This PR is to implement the mark price calculation and apply on the calculation of position value and liquidation.

---
TODO:

- [x] Update changelog

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1203306706086173

tailingchen committed:
feat: add getMarkPrice in AccountBalance

tailingchen committed:
feat: get market price and market twap

daidoli committed:
implement mark price

Co-authored-by: Miya Chen <tailingchen@users.noreply.github.com>

daidoli committed:
add findMedianOfThree

Co-authored-by: Miya Chen <tailingchen@users.noreply.github.com>

daidoli committed:
add _marketTwapInterval

tailingchen committed:
feat: add PerpMathTest in foundry

Co-authored-by: Davis Dai <daidoli@gmail.com>

daidoli committed:
test AccountBalance.getMarkPrice

Co-authored-by: Miya Chen <tailingchen@users.noreply.github.com>

tailingchen committed:
bypass create orderbook

tailingchen committed:
test: fix foundry import

tailingchen committed:
Change intervals of market twap and moving avg

tailingchen committed:
add test to getMarkPrice

Co-authored-by: Davis Dai <daidoli@gmail.com>

daidoli committed:
test moving average mark price

Co-authored-by: Miya Chen <tailingchen@users.noreply.github.com>

tailingchen committed:
Add test to getMarkPrice when marketRegistry is set

tailingchen committed:
chore: add todo comments

daidoli committed:
return index twap if market is not open

Co-authored-by: Miya Chen <tailingchen@users.noreply.github.com>

daidoli committed:
test setMarketRegistry should emit event

Co-authored-by: Miya Chen <tailingchen@users.noreply.github.com>

daidoli committed:
replace get index price in CH with get mark price

Co-authored-by: Miya Chen <tailingchen@users.noreply.github.com>

daidoli committed:
fix AccountBalance.getPositionValue.test.ts

Co-authored-by: Miya Chen <tailingchen@users.noreply.github.com>

daidoli committed:
fix test/accountBalance/AccountBalance.spec.ts

daidoli committed:
fix test/exchange/Exchange.test.ts

tailingchen committed:
feat: overwrite _getMarkPrice for testing purpose

tailingchen committed:
test: fix Exchange.test.ts

tailingchen committed:
test: fix Vault.freeCollateral.test.ts

tailingchen committed:
test: fix Vault.liquidate.test.ts

tailingchen committed:
test: forward block timestamp after init market for calculating twap

tailingchen committed:
test: fix Vault.liquidationGetter.test.ts

tailingchen committed:
test: fix Vault.settleBadDebt.test.ts

tailingchen committed:
test: fix Vault.test.ts

tailingchen committed:
test: fix Vault.withdraw.test.ts

tailingchen committed:
refact: move syncIndexToMarketPrice, syncMarkPriceToMarketPrice to test/shared/utilities.ts

tailingchen committed:
refact: mock index by syncIndexToMarketPrice

tailingchen committed:
test: fix ClearingHouse.accounting.liquidity.test.ts

tailingchen committed:
test: fix ClearingHouse.accounting.xyk.test.ts

tailingchen committed:
test: fix ClearingHouse.badDebt.test.ts

tailingchen committed:
test: fix ClearingHouse.cancelExecessOrders.test.ts

tailingchen committed:
test: fix ClearingHouse.getTotalUnrealizedPnl&getAccountValue.test.ts

tailingchen committed:
rename test file to prevent use '&' symbol in file name

tailingchen committed:
test: fix ClearingHouse.liquidate.maker.ts

tailingchen committed:
test: fix ClearingHouse.newMarket.ts

tailingchen committed:
test: fix ClearingHouse.openPosition.test.ts (wip)

daidoli committed:
fix test case: add other market liquidity below the current tick

Co-authored-by: Miya Chen <miyatlchen@gmail.com>

tailingchen committed:
test: fix ClearingHouse.softCircuitBreak.test.ts

tailingchen committed:
test: fix ClearingHouse.stopMarket.test.ts

tailingchen committed:
test: fix ClearingHouse.takeOver.settleBadDebt.test.ts

tailingchen committed:
test: fix /ClearingHouse.takeOver.test.ts

tailingchen committed:
test: fix ClearingHouse.withdraw.test.ts

tailingchen committed:
test: remove redundant block timestamp forward

tailingchen committed:
test: fix initiateBothTimestamps

tailingchen committed:
test: fix forward block timestamp in initMarket

tailingchen committed:
test: fix funding-induced liquidation to avoid long run

tailingchen committed:
Merge branch 'main' into feature/mark-price

tailingchen committed:
test: fix flaky on "taker has positive pnl on closed market"

tailingchen committed:
refact: refact findMedianOfThree for gas saving (1040->531)

tailingchen committed:
refact: rename _movingAverage* to _premium*

twxia commented on contracts/AccountBalance.sol:
```diff
@@ -512,6 +532,36 @@ contract AccountBalance is IAccountBalance, BlockContext, ClearingHouseCallee, A
+    function _getMarkPrice(address baseToken) internal view virtual returns (uint256) {
+        // Use index twap:
+        //   1. For backward compatible, returns index twap when not switch to mark price yet.
+        //   2. For paused market, returns index twap as mark price.
```
We probabbly don't need 2. 

Both index and mark price won't influnce users after a market gets paused. However, for UX perspective, by using mark price to calculate the unrealized PnL is more appropriate.

tailingchen replied:
When we pause the market, we use index twap as the proposal price for closing. I think we should use the price to calculate the margin for each account.
Ref: 
https://github.com/perpetual-protocol/perp-curie-contract/blob/main/contracts/BaseToken.sol#L44
https://github.com/perpetual-protocol/perp-curie-contract/blob/main/contracts/BaseToken.sol#L138

twxia replied:
Got it, that value seems like a temporary value. For our latest SOP, it seems that we don't need that `_pausedIndexPrice` logic . Probabbly we need to update that part of logic as well.

tailingchen replied:
If we have other price proposal for pause market, I think we should have another PR to refine the pause market process.
Want to only focus mark price change in this PR based on current pause market logic.

tienshaoku commented on test/accountBalance/AccountBalance.getPositionValue.test.ts:
```diff
@@ -308,27 +307,28 @@ describe("AccountBalance.getTotalPositionValue", () => {
         )
-        // 152.711203 * -1.465065799750044640 = -223.731960754
+
+        // 152.711203175725718234 * -1.465065799750044640 = -223.7319610114
```
Need some help: not sure where causes the precision to change

tailingchen replied:
Position value is calculated by mark price instead of by index twap. We print the mark price through console, and the value here is 152.711203175725718234. It's different from the index price above `152.711203`.

tienshaoku replied:
Got it!

twxia replied:
what if we write it somewhat like above section, use `const markPrice = await accountBalance.getMarkPrice(baseToken.address)` to caluclate account value?

twxia commented on contracts/AccountBalance.sol:
```diff
@@ -512,6 +532,36 @@ contract AccountBalance is IAccountBalance, BlockContext, ClearingHouseCallee, A
+            _getMarketPrice(baseToken, _premiumInterval).toInt256().sub(
+                _getIndexPrice(baseToken, _premiumInterval).toInt256()
+            );
+        uint256 indexPremium = _getIndexPrice(baseToken, 0).toInt256().add(premium).toUint256();
```
What do you think that we change `indexPremium` to `fairPrice` or something else? Because premium seems like the difference of two prices

tailingchen replied:
How about `indexWithPremium`? 
I think it's not proper to use `fairPrice` to represent index + premium. The `fairPrice` we propose is the `markPrice` here.

twxia replied:
Fair enough 👍

tailingchen replied:
fixed in aeca2b6a1cca3676588c552b001eabf9b4fde3c2

twxia commented on contracts/AccountBalance.sol:
```diff
@@ -512,6 +532,36 @@ contract AccountBalance is IAccountBalance, BlockContext, ClearingHouseCallee, A
+    }
+
+    function _getIndexPrice(address baseToken, uint32 twapInterval) internal view returns (uint256) {
+        return IIndexPrice(baseToken).getIndexPrice(twapInterval);
```
Do you know the reason why we have `IIndexPrice` and `IBaseToken`?

tailingchen replied:
I'm not sure what's the initial purpose to separate two interfaces. I think we can merge IIndexPrice into IBaseToken.
@Wraecca do you know this?

Wraecca replied:
sorry didn't setup my notification properly after migrating to this repo

in the beginning IBaseToken did not expose to AccountBalance
after v1.1.0 ( close market ) AccountBalance start to access IBaseToken, so it's fine to use IBaseToken instead of IIndexPrice in AccountBalance now

tienshaoku commented on test/clearingHouse/ClearingHouse.accounting.xyk.test.ts:
```diff
@@ -571,19 +568,21 @@ describe("ClearingHouse accounting verification in xyk pool", () => {
 
+            await accountBalance.mockMarkPrice(baseToken.address, parseEther("4"))
             // set index price to let taker pay funding fee
             mockedBaseAggregator.smocked.latestRoundData.will.return.with(async () => {
```
I think this line can be deleted, right?

tailingchen replied:
Actually we can't. Mocking mark price here is to make account value to become under margin, and mocking index price here is to fix the value of founding payment.

tienshaoku commented on test/clearingHouse/ClearingHouse.badDebt.test.ts:
```diff
@@ -56,30 +61,44 @@ describe("ClearingHouse badDebt", () => {
-
-                expect(await clearingHouse.getAccountValue(bob.address)).to.be.eq("111974414000000000000")
+                // bob's account value: 100 + 7.866265610482054835 * 103.199929556499904188 - 800 = 111.7980568745
+                expect(await clearingHouse.getAccountValue(bob.address)).to.be.gt("0")
```
If we know the exact number, suggest expecting it instead of using `gt`

tailingchen replied:
fixed in 558a8d2db7c402d809df779c2a463cd5def4c129

tienshaoku commented on test/clearingHouse/ClearingHouse.badDebt.test.ts:
```diff
@@ -94,25 +113,26 @@ describe("ClearingHouse badDebt", () => {
-                // bob's realized PnL: 0.083 - 0.1/7.866 * 800 = -10.087
-                // bob's account value: 100 + 7.766 * 103.222 - 800 + 0.083 = 101.705 (no bad debt)
-                // bob's free collateral: 100 - 10.087 - (800 - 0.083 + 10.087) * 10% = 8.9126 > 0
+                // bob's realized PnL: 0.083 - 0.05/7.866 * 800 = -10.087
```
Why is here `0.05/7.866` instead of the original `0.1/7.866`?
Isn't the param `0.1` in line 136?

tailingchen replied:
fixed in 558a8d2db7c402d809df779c2a463cd5def4c129

tienshaoku commented on test/clearingHouse/ClearingHouse.newMarket.ts:
```diff
@@ -238,27 +243,28 @@ describe("ClearingHouse new market listing", () => {
-            await mintAndDeposit(fixture, davis, 10000)
+            // drop baseToken mark price
+            await b2qExactOutput(fixture, alice, 10000, baseToken.address)
+            // await b2qExactOutput(fixture, alice, 10000, baseToken3.address)
```
Can be deleted

tienshaoku commented on test/clearingHouse/ClearingHouse.newMarket.ts:
```diff
@@ -238,27 +243,28 @@ describe("ClearingHouse new market listing", () => {
+            await b2qExactOutput(fixture, alice, 10000, baseToken.address)
+            // await b2qExactOutput(fixture, alice, 10000, baseToken3.address)
+            await forwardBothTimestamps(clearingHouse, 1800)
+            await closePosition(fixture, alice, 0, baseToken.address)
```
Is this something we'd like to test? 
If yes, should be in a standalone case instead of here in `beforeEach`.

tailingchen replied:
They are pre-work for this test case. I refine the test case and add more comments for describing the test scenario.
Fixed in e0f729d8ac3d253d6fd6c9dae864242370691a13

tienshaoku commented on contracts/AccountBalance.sol:
```diff
@@ -36,6 +38,9 @@ contract AccountBalance is IAccountBalance, BlockContext, ClearingHouseCallee, A
 
     uint256 internal constant _DUST = 10 wei;
     uint256 internal constant _MIN_PARTIAL_LIQUIDATE_POSITION_VALUE = 100e18 wei; // 100 USD in decimal 18
+    // TODO: might need to move to storage and can set on-demand
```
Are we planning to do this TODO?

tailingchen replied:
As discussed, all protocol configs should defined in ClearingHouseConfig, MarketRegistry and CollateralManager.
Fixed in another PR https://github.com/perpetual-protocol/perp-curie-contract/pull/37

twxia commented on test/foundry/perpMath/PerpMath.t.sol:
```diff
@@ -0,0 +1,13 @@
+    function test_findMedianOfThree() public {
+        assertEq(PerpMath.findMedianOfThree(1, 2, 3), uint256(2));
+        assertEq(PerpMath.findMedianOfThree(2, 1, 3), uint256(2));
+        assertEq(PerpMath.findMedianOfThree(1, 3, 2), uint256(2));
```
I think below cases should be tested as well:
2 3 1
3 1 2
3 2 1
above are permutations (including L9-L11)

below make sure that in the duplication situaiton, the result is what we want
1 1 1
1 2 2

tailingchen replied:
fixed in f2db4928c431c5a838bfe1a3f224c510eb418749

tienshaoku commented on test/helper/marketHelper.ts:
```diff
@@ -34,6 +38,13 @@ export async function initMarket(
 
+    if (isForwardTimestamp) {
+        const clearingHouse = fixture.clearingHouse as TestClearingHouse
+        await initiateBothTimestamps(clearingHouse)
```
Two questions
1. Not sure why we add this logic here cuz I'm still seeing `initiateBothTimestamps` in such as `ClearingHouse.badDebt.test.ts`. Do we simply forget to remove it?
2. Since we `initiateBothTimestamps` already, I think we don't need to `forwardBothTimestamps` again, right?

tailingchen replied:
1. No. It's required.
In most test cases, we need to deploy market at the beginning of beforeEach  and setup some pre-work for each test case by sending couples of transactions.  Due to we need to calculate mark price, so we initial and forward both of bloc timestamps in `initMarket` function. But in some tests for funding payment, the values of mock block timestamp and evm block timestamp are gradually separated after some pre-work transactions (ex: adding liquidity), so it's better to initial both block timestamps again to make funding payment test easier and prevent flaky funding payment tests.

2. forwardBothTimestamps is for mark price calculation.

tienshaoku replied:
Not getting your point haha
Will gather you when you're available!

tailingchen replied:
After discussed, will use forwardRealTimestamp to prevent `OLD` reverted by Uniswap contract while mark price calculation.
Fixed in be6bfb5d26faf4773183824cf927fe693557acf7

twxia commented on contracts/interface/IAccountBalance.sol:
```diff
@@ -232,4 +239,9 @@ interface IAccountBalance {
         address baseToken,
         int256 accountValue
     ) external view returns (int256);
+
```
L221 should be updated as well.

`getTotalPositionValue `'s comment
```
/// @dev we use `15 mins` twap to calc position value
```

tailingchen replied:
Fixed in 01ba903afff31d0aef7bb7f137760ef0241f59fc

twxia commented on test/clearingHouse/ClearingHouse.accounting.xyk.test.ts:
```diff
@@ -129,11 +127,12 @@ describe("ClearingHouse accounting verification in xyk pool", () => {
         await collateral.mint(insuranceFund.address, parseUnits("1000000", 6))
+
+        // initiate both the real and mocked timestamps to enable hard-coded funding related numbers
+        // NOTE: Should be the last step in beforeEach
```
Not quite understand why we move it to the bottom. What's the purpose of this?

tailingchen replied:
We have two block timestamps in test. One is mocked block timestamp, which is for founding payment testing purpose. Another one is evm block timestamp.

To prevent flaky on funding payment tests, it's better to sync mocked block timestamp and evm block timestamp before running each test cases. That's why it should be placed at the bottom of beforeEach. To prevent other devs change the order, add the NOTE comments here for reminder.

twxia commented on test/clearingHouse/ClearingHouse.accounting.xyk.test.ts:
```diff
@@ -571,19 +568,21 @@ describe("ClearingHouse accounting verification in xyk pool", () => {
-            while ((await clearingHouse.getAccountValue(taker.address)).gt(0)) {
-                await forwardBothTimestamps(clearingHouse, 3000)
+            // taker is not bankruptcy yet, even he has loss
+            expect(await clearingHouse.getAccountValue(taker.address)).to.be.gt(0)
```
account value > 0 might still be liquidatable. probabbly needs to use `_isLiquidatable` 's logic to make sure that this account right now is healthy.

tailingchen replied:
Fixed in 2fa7e1f6904d49042e96ef6d89c54adc43c15334

twxia commented on test/clearingHouse/ClearingHouse.badDebt.test.ts:
```diff
@@ -56,30 +61,44 @@ describe("ClearingHouse badDebt", () => {
                 await q2bExactInput(fixture, bob, "800", baseToken.address)
-                // market price = index price = 103.222
+
+                // To ignore founding payment
```
```suggestion
                // To ignore funding payment
```

tailingchen replied:
fixed in 558a8d2db7c402d809df779c2a463cd5def4c129

tailingchen committed:
test: assert exact number & fix comments

tailingchen committed:
refact: rename indexPremium to indexWithPremium

twxia commented on test/clearingHouse/ClearingHouse.badDebt.test.ts:
```diff
@@ -56,30 +61,44 @@ describe("ClearingHouse badDebt", () => {
         await collateral.mint(bob.address, parseUnits("100", decimals))
         await deposit(bob, vault, 100, collateral)
 
+        // prepare collateral for carol
```
It seems that we still can use Alice, no need Carol, right?

tailingchen replied:
Yes, we can use Alice here. But since the purpose of this test case is not to test that the taker and maker are the same person,  to make tests clearer and no confusion, we specify Alice is maker, and Bob and Carol are takers.

tienshaoku commented on test/clearingHouse/ClearingHouse.openPosition.test.ts:
```diff
@@ -1351,17 +1362,24 @@ describe("ClearingHouse openPosition", () => {
+                //     marketPrice: '146.88860454062238813'
+                // }
+
+                // realizedPnL = -0.000001209503724417
```
Can you help indicate that this `realizedPnL` is `fundingPayment`? That'd help with understanding!

tailingchen replied:
add more comments. fixed in 825413845711cca44dcb0c51a361f7d7c2f70dd6

tailingchen committed:
refact: remove duplicate import and reorder import statements

twxia commented on test/clearingHouse/ClearingHouse.softCircuitBreak.test.ts:
```diff
@@ -127,13 +111,10 @@ describe("ClearingHouse softCircuitBreak", () => {
             await mintAndDeposit(fixture, davis, 10000)
-
-            // increase blockTimestamp
-            await clearingHouse.setBlockTimestamp(blockTimeStamp + 1)
```
what's the reason that we don't need this line?

tailingchen replied:
I think it's redundant code statement. In these test case, should do liquidate even the block timestamp is not forward.

tailingchen committed:
test: remove useless comment

twxia commented on test/foundry/helper/Setup.sol:
```diff
@@ -13,11 +13,13 @@ import "../../../contracts/QuoteToken.sol";
-import "@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol";
-import "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+import { IPriceFeed } from "@perp/perp-oracle-contract/contracts/interface/IPriceFeed.sol";
+// import { UniswapV3Factory } from "@uniswap/v3-core/contracts/UniswapV3Factory.sol";
+// import { UniswapV3Pool } from "@uniswap/v3-core/contracts/UniswapV3Pool.sol";
```
```suggestion
```

tailingchen replied:
fixed in 01ba903afff31d0aef7bb7f137760ef0241f59fc

twxia commented on test/vault/Vault.test.ts:
```diff
@@ -317,6 +317,7 @@ describe("Vault test", () => {
                 mockedBaseAggregator.smocked.latestRoundData.will.return.with(async () => {
                     return [0, parseUnits("150", 6), 0, 0, 0]
                 })
+                await accountBalance.mockMarkPrice(baseToken.address, parseEther("150"))
```
In this file, what's the reason that we didn't remove the mock of index price? It seems that we don't need it.

tailingchen replied:
No, we can't.
Mocking mark price is for account value calculation, and mocking index price is for funding payment calculation.
Add more comments to describe why we need to mock mark price and index price in tests.
Fixed in 847555cfb33bae9c6fe9c1e162a9db538458b735

twxia commented on test/clearingHouse/ClearingHouse.stopMarket.test.ts:
```diff
@@ -222,11 +223,13 @@ describe("Clearinghouse StopMarket", async () => {
 
                 // set a lower price to have loss on baseToken2 market
-                mockedBaseAggregator2.smocked.latestRoundData.will.return.with(async () => {
+                mockedBaseAggregator.smocked.latestRoundData.will.return.with(async () => {
```
```suggestion
                mockedBaseAggregator2.smocked.latestRoundData.will.return.with(async () => {
```

tailingchen replied:
Fixed in 7b8a7a3ea400702a74b58f1594349c410ca28a00

tailingchen committed:
refact: refact test case and add more comments to describe test scenario

tailingchen committed:
test: add more test cases for PerpMath

tailingchen committed:
refact: update comment

tailingchen committed:
test: fix tests and add more comments to describe test scenario

tailingchen committed:
fix: refine test case

tailingchen committed:
refact: refine the comments

tailingchen committed:
refact: refine test case and remove unused function

tailingchen committed:
refact: describe why we need to mock mark price and index price in tests

tailingchen committed:
feat: move mark price params from constant to contract storage

tailingchen committed:
refact:
  1. add default values of intervals used in mark price in initialize
  2. add function _isEnableMarkPrice to check if mark price params are set already
  3. refine nespec for getMarkPrice function

tailingchen committed:
test: tests on mark price is not enabled

twxia commented on test/foundry/accountBalance/accountBalance.t.sol:
```diff
@@ -0,0 +1,147 @@
+        secondsAgos[0] = _marketTwapInterval;
+        secondsAgos[1] = 0;
+
+        // Price: |---- 90 (10min)---|---- 95 (10min)---|---- 100 (10min)---|
+        // Tick:  |-- 45000 (10min)--|-- 45541 (10min)--|-- 46054 (10min) --|
+        int56[] memory tickCumulatives = new int56[](2);
+        tickCumulatives[0] = 0;
+        tickCumulatives[1] = 81957000;
```
Could we just directly mockcall `getSqrtMarkTwapX96`? so we don't need to test library logic which is out of scope.

tailingchen replied:
UniswapV3Broker.getSqrtMarkTwapX96 is an internal function.
vm. mockCall does not currently work on internal calls. See issue [#432](https://github.com/foundry-rs/foundry/issues/432).

tailingchen committed:
test: use forwardRealTimestamp in initMarket

tailingchen committed:
test: add foundry tests for clearingHouseConfig setters

tailingchen committed:
test: fix tests

tailingchen committed:
refact: get intervals of market twap and premium through one getter

tailingchen committed:
Update CHANGELOG.md

tailingchen committed:
Merge branch 'feature/mark-price' into feature/mark-price-move-config-to-storage

tailingchen committed:
Update CHANGELOG.md

tailingchen committed:
fix typo

tienshaoku committed:
# Avoiding abbreviation.

tienshaoku committed:
# Avoiding s in function name.

tienshaoku committed:
# _isEnableMarkPrice() renamed and refactored.

tienshaoku committed:
# Preparation for changing accountBalance.sol.

tienshaoku committed:
# Foundry test filenames updated.

tienshaoku committed:
# expectEmit() params updated.

tailingchen committed:
Merge pull request #40 from perpetual-protocol/refactor/mark-price-config

Refactor/mark price config

tailingchen committed:
Merge pull request #37 from perpetual-protocol/feature/mark-price-move-config-to-storage

refactor: move mark price config from constant to storage

tienshaoku committed:
# Most tests of ClearingHouseConfig migrated to foundry.

tienshaoku committed:
# Constants extracted to Constant.sol.

tienshaoku committed:
# Events in ClearingHouseConfig moved to IClearingHouseConfigEvent in IClearingHouseConfig.sol.

tienshaoku committed:
BackstopLiquidityProvider deprecated.

tienshaoku committed:
# Param of event TwapIntervalChanged changed back.

tailingchen committed:
Merge pull request #47 from perpetual-protocol/refactor/ClearingHouseConfig-cleanup

refactor/ClearingHouseConfig-cleanup

Wraecca committed:
simplify _getMarketPrice

Wraecca committed:
update accountBalance's interface and storage

Wraecca committed:
fix tests

Wraecca committed:
update changelog

tailingchen committed:
Merge pull request #50 from perpetual-protocol/feature/mark-price-rm-ab-registry

Feature/simplify mark price

tailingchen committed:
Merge pull request #44 from perpetual-protocol/test/migrate-mark-price-related-tests-to-foundry

refactor ClearingHouseConfig

tailingchen committed:
feat: rename markTwap to marketTwap and remove redundant import

tailingchen committed:
feat: fix rename and revert the new function Exchange.getSqrtMarketTwapX96

tailingchen committed:
feat: add Exchange.getSqrtMarketTwapX96 and rename event param from markTwap to marketTwap

tailingchen committed:
feat: update CHANGELOG.md

tailingchen committed:
Merge pull request #54 from perpetual-protocol/feature/mark-price-clarify-variables-naming-and-fix-minors

feat: rename markTwap to marketTwap and remove redundant import

phchan9 commented on contracts/lib/UniswapV3Broker.sol:
```diff
@@ -227,11 +227,11 @@ library UniswapV3Broker {
     ///      as twapInterval is too short and makes getting twap over such a short period meaningless
-    function getSqrtMarkTwapX96(address pool, uint32 twapInterval) internal view returns (uint160) {
-        return getSqrtMarkTwapX96From(pool, 0, twapInterval);
+    function getSqrtMarketTwapX96(address pool, uint32 twapInterval) internal view returns (uint160) {
+        return getSqrtMarketTwapX96From(pool, 0, twapInterval);
```
This is also a breaking change for ppl (like partnership or community members) who already uses this library code but not sure if this is so important that we need to keep **getSqrtMarkTwapX96**  for a while and deprecate it later.

tailingchen replied:
It's a library, not a contract. It's fine even if the partners or other community members continue to use the old UniswapV3Broker library.

phchan9 commented on contracts/lib/UniswapV3Broker.sol:
```diff
@@ -227,11 +227,11 @@ library UniswapV3Broker {
     }
 
-    function getSqrtMarkTwapX96From(
+    function getSqrtMarketTwapX96From(
```
same here as the above comment.

tailingchen replied:
answered above

phchan9 commented on test/clearingHouseConfig/ClearingHouseConfig.spec.ts:
```diff
@@ -1,85 +0,0 @@
-import { ethers, waffle } from "hardhat"
-import { ClearingHouseConfig } from "../../typechain"
-
-describe("ClearingHouseConfig Spec", () => {
```
why we remove this test suit

tailingchen replied:
Already moved to foundry test. https://github.com/perpetual-protocol/perp-curie-contract/pull/30/files#diff-c5ad1c38390945488c5c3035286fd816c3f47bc72ac9cbebd8a3fb11aaf51c81

phchan9 commented on test/vault/Vault.settleBadDebt.test.ts:
```diff
@@ -90,7 +92,6 @@ describe("Vault settleBadDebt (assume zero IF fee)", () => {
         await addOrder(fixture, bob, 500, 1000000, 0, 150000)
 
         // Carol will liquidate Alice's position
-        await clearingHouseConfig.setBackstopLiquidityProvider(carol.address, true)
```
what is the purpose  to remove this line?

tailingchen replied:
Since we already release new liquidation mechanism, the bad debt liquidation can be done by anyone not just configured BLPs. And BackstopLiquidityProvider will be deprecated at this release, check [here](https://github.com/perpetual-protocol/perp-curie-contract/pull/30/files#diff-19ab1b74485b7ebdbed9a7cbec8e79ee43e547847490d52650e06ae8c72f3729L88).

tailingchen committed:
feat: avoid to access uniswap directly

tailingchen committed:
Merge pull request #55 from perpetual-protocol/feature/mark-price-not-import-uniswap-except-exchange-and-orderbook

feat: avoid to access uniswap directly
